### PR TITLE
Products API 

### PR DESCRIPTION
**Pull request by rkistaps**
Added new Products Api endpoint to SDK

Docs: https://www.printful.com/docs/products

### CODE REVIEW

File: src/PrintfulProductsApi.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+use Printful\Structures\Sync\SyncProductUpdateParameters;
+
+/**
+ * Class PrintfulProductsApi
+ *
+ * API Docs: https://www.printful.com/docs/products
+ *
+ * @package Printful
+ */
+class PrintfulProductsApi
```
Comment (gatispelcers) [inline src/PrintfulProductsApi.php]: Varbūt pieturēties pie vienota stila un saukt vienkārši par PrintfulProducts kā citas API endpointus wrapojošās klases?

File: examples/ProductsApi/delete.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+use Printful\Exceptions\PrintfulApiException;
+use Printful\Exceptions\PrintfulException;
+use Printful\PrintfulApiClient;
+use Printful\PrintfulProductsApi;
+
+require_once __DIR__ . '../../vendor/autoload.php';
```
Comment (gatispelcers) [inline examples/ProductsApi/delete.php]: require feilo, jo __DIR__ nav trailing slash. Šis attiecas uz visiem examples failiem, arī tiem, kas nav šajā MR.

Comment (rkistaps) [reply examples/ProductsApi/delete.php]: Saliku trailing slash savos example

File: src/PrintfulProductsApi.php (hunk 1 above)
Comment (rkistaps) [reply src/PrintfulProductsApi.php]: Pārsaukts

File: src/Structures/Sync/Responses/SyncVariantProductResponse.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+class SyncVariantProductResponse
+{
+    /** @var int */
+    public $variantId;
+
+    /** @var int */
+    public $productId;
+
+    /** @var string */
+    public $image;
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Structures/Sync/Responses/SyncVariantProductResponse.php]: Property naming is identical to API response. I suggest to keep it consistent in that way.

File: src/Structures/Sync/Requests/SyncVariantRequestFile.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Printful\Structures\Sync\Requests;
+
+class SyncVariantRequestFile
+{
+    const DEFAULT_TYPE = 'default';
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Structures/Sync/Requests/SyncVariantRequestFile.php]: flipped

File: src/Structures/Sync/Responses/SyncProductResponse.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    public $externalId;
+
+    /** @var string */
+    public $name;
+
+    /** @var int */
+    public $variants = 0;
+
+    /** @var int */
+    public $synced = 0;
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Structures/Sync/Responses/SyncProductResponse.php]: It actually represents number of synced SyncVariants belonging to this product. Naming is consistent with API response. Probably property description is needed.

File: src/Factories/Sync/ParameterFactory.php [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+namespace Printful\Factories\Sync;
+
+use Printful\Exceptions\PrintfulSdkException;
+use Printful\Structures\Sync\Requests\SyncVariantRequestFile;
+use Printful\Structures\Sync\Requests\SyncVariantRequestOption;
+use Printful\Structures\Sync\SyncProductCreationParameters;
+use Printful\Structures\Sync\Requests\SyncVariantRequest;
+use Printful\Structures\Sync\SyncProductUpdateParameters;
+
+class ParameterFactory
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Factories/Sync/ParameterFactory.php]: Done

File: src/Structures/Sync/Requests/SyncProductRequest.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+class SyncProductRequest
+{
+    /** @var string */
+    public $externalId;
+
+    /** @var string */
+    public $name;
+
+    /** @var string */
+    public $thumbnail = '';
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Structures/Sync/Requests/SyncProductRequest.php]: Property naming is identical to API response. I suggest to keep it consistent.

File: src/Structures/Sync/Responses/SyncVariantResponse.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    public $variantId;
+
+    /** @var float */
+    public $retailPrice;
+
+    /** @var string */
+    public $currency;
+
+    /** @var SyncVariantProductResponse */
+    public $product;
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Structures/Sync/Responses/SyncVariantResponse.php]: Property naming is identical to API response. I suggest to keep it consistent. 

Also it isn't even full syncProduct, it's just some trimmed/half-random product data. Refer to API for more insight: https://www.printful.com/docs/products#getSingleSyncVariant

File: tests/Products/UpdateTest.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function testUpdateVariant()
+    {
+        $syncProduct = $this->createProduct();
+        $syncProduct = $this->apiEndpoint->getProduct($syncProduct->id);
+
+        $variant = $syncProduct->syncVariants[0];
+
+        $request = new SyncVariantRequest;
+        $request->externalId = 'my-sdk-updated-external-id-' . rand(1, 10);
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply tests/Products/UpdateTest.php]: Fixed

File: src/Structures/Sync/Responses/SyncVariantFileResponse.php [hunk 10]
```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace Printful\Structures\Sync\Responses;
+
+class SyncVariantFileResponse
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Structures/Sync/Responses/SyncVariantFileResponse.php]: Renamed `\Printful\Structures\Order\OrderItemFile` to `\Printful\Structures\File`. Replaced `SyncVariantFileResponse` with `File` class.

File: tests/Products/ProductsTestBase.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+     * @param bool $useCache
+     * @return SyncProductResponse
+     * @throws \Printful\Exceptions\PrintfulApiException
+     * @throws \Printful\Exceptions\PrintfulException
+     */
+    protected function createProduct($useCache = true)
+    {
+        static $result;
+
+        if ($useCache && $result) {
```
[the remark this answers is not in the document]
Comment (rkistaps) [reply tests/Products/ProductsTestBase.php]: Removed caching

File: src/Factories/Sync/ParameterFactory.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (rkistaps) [reply src/Factories/Sync/ParameterFactory.php]: Probably need another CR

**Comment by graber-1** [comment]
@briedis, @gatispelcers any chance this PR is re-reviewed? I see all comments have replies from the author.

**Comment by briedis** [comment]
@graber-1 Sorry about that. We'll take a look again...

**Comment by graber-1** [comment]
Nice, thanks. I'll need this API so I may take over if the author is not available anymore.

**Comment by briedis** [comment]
@graber-1 Until we release this, you can use the API directly without our SDK classes.

**Comment by sebastian-king** [comment]
What's the delay on getting this PR merged? It's been almost a year since the last update.

**Comment by graber-1** [comment]
Yeah, this is total BS, I created my own class for handling Printful API requests already.

**Comment by rkistaps** [comment]
PR has been merged and a new release tagged. Sorry for the delay.

**Pull request merged**